### DISCUSSION TITLE
Default repo for table (and/or BaseTable)

### DISCUSSION THREAD

**Discussion by IlyaSemenov** [romeerez/orchid-orm#100] [Ideas]
Consider a repo:

```ts
export const userRepo = createRepo(db.user, {
  queryMethods: {
    search(q, lookup: string) { ... }
  },
});
```

This is typed (and apparently implemented) as extension to `db.user`. This code becomes immediately possible:

```ts
userRepo.select(...).where(...).search("foo")
```

I would like to somehow define/provide a default "repo" for `db.user` with general purpose query extensions, so that the following code also becomes possible:

```ts
db.user.select(...).where(...).search("foo")
```

Perhaps something like this?

```ts
class UserTable extends BaseTable {
  readonly table = 'user'

  repo = this.createRepo(t => ({
    queryMethods: {
      search(q, lookup: string) { ... }
    },	
  }))

  // or:
  queryMethods = this.setQueryMethods(t => ({ search: ... }))
}
```

Ideally, I would like the same to be possible on `BaseTable` itself, as I do have need for generic methods applied to all models (e.g. shortcuts for working with "pseudo-deleted" objects which are marked in database instead of the actual deletion). That was easily possible in Objection. I do realise that there are serious typings implications so that is a different request. However, per-table repo injection seems to be a relatively easy addition?

**Comment by romeerez** [comment]
The ORM is partially inspired by ActiveRecord from ruby, where this is possible and we did such queries all the time. As a downside, the models were enormous, they contained validation, business logic, callbacks, helpers, methods for single records, and methods for collections. And what I did about this is that I tried to keep it in different files, like `SearchMethods` contained custom methods for the model only for the search.

So even when it's supported, it's not always for the better. But here in TS it's too problematic. `db` depends on tables, so the tables cannot use the `db`, and so one table cannot access `db.otherTable`, nor it cannot access its own relations to query them. In the table class, relations are just "thunks"  that are processed later when calling `createORM`.

What you are suggesting is probably possible, but only if you won't access any relations in the repo, and it may add some problems with TS, because `db.user` will be a complete user with all its relations, but in the repo it will be a limited version of `db.user` - two different types.

**Reply by IlyaSemenov** [reply]
Thank you for the thorough reply. These are valid points so perhaps I retract my idea for now.

**Reply by IlyaSemenov** [reply]
I have a somewhat unusual setup where I provide authentication modifiers to each table, and for now I ended up creating a separate repo for each of the tables, with all of these repos exposing the same method with slightly different implementations.

I was exploring options on how to reduce boilerplate for this scenario, in particular having to duplicate imports (`db.user.all()` vs `restricted_db.user.myrestrict(context).all()`)

**Reply by romeerez** [reply]
The only idea I have for now for the default repos would be to create a `repos` and use it instead of `db`:

```ts
export const db = orchidORM(dbConfig, {
  tomato: TomatoTable,
  pumpkin: PumpkinTable,
  cherry: CherryTable,
})

// must be in a different file because of cyclic imports:
export const repos = {
  // db.tomato is a default, so no need to pass:
  tomato: tomatoRepo,
  pumpkin: pumpkinRepo,
  
  // cherry doesn't have a repo, so it's db.cherry
  cherry: db.cherry,
}

// and in the code to use repos instead
repos.tomato.find(123)
repos.pumpkin.where(...)
```

Maybe this won't be very helpful, just as an option.

**Reply by IlyaSemenov** [reply]
Right, that works. Just somehow feels not super elegant... and probably not worth it. Keeping a separate collection of repos makes it more explicit in code. Besides, only part of the code (direct user-facing) needs to work with authorizing repos, other backend services don't need it. Anyhow, thanks for feedback. I think in the end I'm good with what we have now.

**Comment by romeerez** [comment]
Are you familiar with some ORM for a compiled language, C# or Java? Wondering if they can handle this.
